File: ios/LivtetJigsaw/Sources/LivtetJigsaw/Model/Validator.swift
import Foundation

public typealias ValidationListener = () -> Void

public protocol Validator: AnyObject {
    var valid: Bool { get set }
    func isValid(_ puzzle: Puzzle) -> Bool
    func validate(_ puzzle: Puzzle)
    func onValid(_ f: @escaping ValidationListener)
}

public final class NullValidator: Validator {
    public var valid: Bool = false
    private var listeners: [ValidationListener] = []

    public init() {}

    public func isValid(_ puzzle: Puzzle) -> Bool { false }

    public func validate(_ puzzle: Puzzle) {}

    public func onValid(_ f: @escaping ValidationListener) { listeners.append(f) }
}

public final class PieceValidator: Validator {
    public var valid: Bool = false
    private var listeners: [ValidationListener] = []
    private let check: (Piece) -> Bool

    public init(_ check: @escaping (Piece) -> Bool) { self.check = check }

    public func isValid(_ puzzle: Puzzle) -> Bool {
        puzzle.pieces.allSatisfy(check)
    }

    public func validate(_ puzzle: Puzzle) {
        let wasValid = valid
        valid = puzzle.pieces.allSatisfy(check)
        if !wasValid && valid { listeners.forEach { $0() } }
    }

    public func onValid(_ f: @escaping ValidationListener) { listeners.append(f) }
}

public final class PuzzleValidator: Validator {
    public var valid: Bool = false
    private var listeners: [ValidationListener] = []
    private let check: (Puzzle) -> Bool

    public init(_ check: @escaping (Puzzle) -> Bool) { self.check = check }

    public func isValid(_ puzzle: Puzzle) -> Bool { check(puzzle) }

    public func validate(_ puzzle: Puzzle) {
        let wasValid = valid
        valid = check(puzzle)
        if !wasValid && valid { listeners.forEach { $0() } }
    }

    public func onValid(_ f: @escaping ValidationListener) { listeners.append(f) }

    public static func relativeRefs(_ expected: [[CGFloat]]) -> (Puzzle) -> Bool {
        return { puzzle in
            guard puzzle.refs.count == expected.count else { return false }
            return zip(puzzle.refs, expected).allSatisfy { ref, exp in
                guard ref.count == 2, exp.count == 2 else { return false }
                return abs(ref[0] - exp[0]) < 0.01 && abs(ref[1] - exp[1]) < 0.01
            }
        }
    }
}

File: android/jigsaw/src/main/kotlin/net/olamaelcu/livtet/jigsaw/model/Validator.kt
package net.olamaelcu.livtet.jigsaw.model

typealias ValidationListener = (Puzzle) -> Unit

abstract class Validator {
    private val listeners = mutableListOf<ValidationListener>()
    internal var _valid: Boolean? = null
    open val valid get() = _valid ?: false
    open val isNull get() = false
    abstract fun isValid(puzzle: Puzzle): Boolean
    fun validate(puzzle: Puzzle) { val was = _valid; updateValidity(puzzle); if(_valid == true && was != true) fireValid(puzzle) }
    fun updateValidity(puzzle: Puzzle) { _valid = isValid(puzzle) }
    fun fireValid(puzzle: Puzzle) { listeners.forEach { it(puzzle) } }
    fun onValid(f: ValidationListener) { listeners.add(f) }
}

class NullValidator : Validator() { override val isNull=true; override fun isValid(p: Puzzle)=false }
class PieceValidator(val condition: (Piece) -> Boolean) : Validator() { override fun isValid(p: Puzzle) = p.pieces.all { condition(it) } }
class PuzzleValidator(val condition: (Puzzle) -> Boolean) : Validator() { override fun isValid(p: Puzzle) = condition(p) }
